### PR TITLE
[iOS] Fix SwipeView stays open on iOS after updating content

### LINKED ISSUE #19541: [iOS] - Swipeview with collectionview issue
### Description

Hi,
in a Collectionview if there is a swipeview opened and the list binded at the collectionview is refreshed, the swipe is not resetted.

On android works.

### Steps to Reproduce

1. create a collectionview with swipeview that contains buttons
2. create a timer that refresh the list
3. perform a swipe
4. wait till the refresh
5. the swipe are not reset and the button are put on another random list element

### Link to public reproduction project repository

https://github.com/MattePozzy/MauiSwipeViewIssue

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

_No response_

### Did you find any workaround?

No

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by devanathan-vaithiyanathan**
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

### Issue Details
On iOS, when a SwipeView is placed inside a CollectionView and its content is updated, the swipe items remain in the open state. 

### Description of Change

When the BindingContext changes, manually close the open state of the SwipeView.

### Issues Fixed

Fixes #19541 

**Tested the behavior in the following platforms.**
- [x] Android
- [x] Windows
- [x] iOS
- [x] Mac

| Before  | After  |
|---------|--------|
| **iOS**<br> <video src="https://github.com/user-attachments/assets/75a1ec0f-faec-4f7e-9093-72298f1d63c4" width="300" height="600"> | **iOS**<br> <video src="https://github.com/user-attachments/assets/847b5c7e-9339-4749-aae7-a8e0fb73baed" width="300" height="600"> |

### CODE REVIEW

File: src/Controls/src/Core/SwipeView/SwipeView.cs [hunk 1]
```diff
@@ -128,6 +128,16 @@ protected override void OnBindingContextChanged()
 			SetInheritedBindingContext(RightItems, BindingContext);
 			SetInheritedBindingContext(TopItems, BindingContext);
 			SetInheritedBindingContext(BottomItems, BindingContext);
+
+			// In Windows and Android, new cells are created when rebinding, so _isOpen is false.
+			// In iOS, CollectionView reuses cells, so SwipeView remain open (_isOpen = true).
+			// When new cells are recreated, the swipe state is reset. 
+			// When cells are reused (iOS), the open state is maintained.
+			// Fix for iOS: Close SwipeView when BindingContext changes to prevent stale open state
+			if (_isOpen)
```
Comment (jsuarezruiz) [inline src/Controls/src/Core/SwipeView/SwipeView.cs:138]: Should use a compilation directive and only do this on iOS?

Comment (jsuarezruiz) [reply src/Controls/src/Core/SwipeView/SwipeView.cs:138]: And Catalyst

Comment (devanathan-vaithiyanathan) [reply src/Controls/src/Core/SwipeView/SwipeView.cs:138]: @jsuarezruiz , Based on suggestion, I have modified the changes.

**Comment by rmarinho** [comment]
## 🤖 AI Summary


<details>
<summary>📊 <strong>Expand Full Review</strong></summary>

---

<details>
<summary><strong>🔍 Pre-Flight — Context & Validation</strong></summary>

---

<details>
<summary>📝 <strong>Review Session</strong> — <strong>condition added</strong> · <a href="https://github.com/dotnet/maui/commit/fc467b1c507385c2d97ccd963abbc55bff566062"><code>fc467b1</code></a></summary>

---

**Issue:** #19541 - [iOS] SwipeView with CollectionView issue
**PR:** #31248 - [iOS] Fix SwipeView stays open on iOS after updating content
**Platforms Affected:** iOS (primary), MacCatalyst
**Files Changed:** 1 fix file, 3 snapshot files, 2 test files

### Issue Summary
On iOS, when a SwipeView is open inside a CollectionView and the CollectionView's `ItemsSource` is refreshed, the SwipeView remains in the open state. This happens because iOS/MacCatalyst CollectionViews **reuse cells** (unlike Android/Windows which create new cells), so the old swipe open state persists after rebinding. On Android, this works correctly because new cells are created on rebind.

### Fix Description
In `SwipeView.OnBindingContextChanged()`, close the SwipeView when `_isOpen == true`, but only on iOS/MacCatalyst (using `#if IOS || MACCATALYST` compiler directive). This prevents the stale open state when cells are recycled.

### Files Changed

| File | Type | Description |
|------|------|-------------|
| `src/Controls/src/Core/SwipeView/SwipeView.cs` | Fix | Add close logic in `OnBindingContextChanged` for iOS/MacCatalyst |
| `src/Controls/tests/TestCases.HostApp/Issues/Issue19541.cs` | Test | Host app reproduction page |
| `src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue19541.cs` | Test | NUnit UI test |
| `src/Controls/tests/TestCases.iOS.Tests/snapshots/ios/SwipeItemShouldBeCloseWhenUpdateTheCollectionView.png` | Snapshot | iOS screenshot baseline |
| `src/Controls/tests/TestCases.Android.Tests/snapshots/android/SwipeItemShouldBeCloseWhenUpdateTheCollectionView.png` | Snapshot | Android screenshot baseline |
| `src/Controls/tests/TestCases.Mac.Tests/snapshots/mac/SwipeItemShouldBeCloseWhenUpdateTheCollectionView.png` | Snapshot | macOS screenshot baseline |

### PR Discussion Summary

| File:Line | Reviewer | Comment | Status |
|-----------|----------|---------|--------|
|  | jsuarezruiz | Should use a compilation directive and only do this on iOS? And Catalyst |  RESOLVED (author implemented `#if IOS \|\| MACCATALYST`) |

No other review disagreements or edge cases noted. Azure Pipelines UI tests were run. PR was rebased on 2025-10-10.

### Fix Candidates
| # | Source | Approach | Test Result | Files Changed | Notes |
|---|--------|----------|-------------|---------------|-------|
| PR | PR #31248 | Close SwipeView in `OnBindingContextChanged` for iOS/MacCatalyst when `_isOpen ==    PENDING (Gate) | `SwipeView.cs` (+12) | Original PR |true` |

</details>

</details>

---

<details>
<summary><strong>🚦 Gate — Test Verification</strong></summary>

---

<details>
<summary>📝 <strong>Review Session</strong> — <strong>condition added</strong> · <a href="https://github.com/dotnet/maui/commit/fc467b1c507385c2d97ccd963abbc55bff566062"><code>fc467b1</code></a></summary>

---

**Result PASSED:** 
**Platform:** iOS (iPhone 11 Pro simulator)
**Mode:** Full Verification

- Tests FAIL without fix (bug detected correctly) 
- Tests PASS with fix (fix validated) 

**Test:** `SwipeItemShouldBeCloseWhenUpdateTheCollectionView` in `Issue19541`
**Fix files reverted for test-without-fix run:** `src/Controls/src/Core/SwipeView/SwipeView.cs`

</details>

</details>

---

<details>
<summary><strong>🔧 Fix — Analysis & Comparison</strong></summary>

---

<details>
<summary>📝 <strong>Review Session</strong> — <strong>condition added</strong> · <a href="https://github.com/dotnet/maui/commit/fc467b1c507385c2d97ccd963abbc55bff566062"><code>fc467b1</code></a></summary>

---

### Fix Candidates

| # | Source | Approach | Test Result | Files Changed | Notes |
|---|--------|----------|-------------|---------------|-------|
| 1 | try-fix | Add IsOpen property mapper on iOS handler; use  in handler; set  in  |  PASS | 3 files + 2 API files | More complex; requires new public API surface |
| 2 | try-fix | Unconditionally set  and call  in  (no  guard) |  PASS | 1 file | Slight risk: fires even when not open |
| 3 | try-fix | Listen for parent CollectionView  PropertyChanged and call  when open |  PASS | 1 file | Targets refresh path directly; couples to CollectionView type |
| 4 | try-fix | Close SwipeView when Parent becomes null in  (cell return to reuse pool) |  PASS | 1 file | Too broad; fires on any parent detachment, not just ItemsSource refresh |
| PR | PR #31248 | In , close with  guard +  check via  |  PASS (Gate) | 1 file (+12 lines) | Most minimal and targeted |

**Exhausted:** Yes (cross-pollination: all 4 models say NO NEW IDEAS after Round 1)
**Selected Fix:** PR's  It is the most minimal change (single file, 12 lines), correctly platform-gated to iOS/MacCatalyst using compiler directives, guards against unnecessary calls with `if (_isOpen)`, and uses the established `ISwipeView` interface for requesting close. All alternatives add complexity or have unintended side effects.fix

</details>

</details>

---

<details>
<summary><strong>📋 Report — Final Recommendation</strong></summary>

---

<details>
<summary>📝 <strong>Review Session</strong> — <strong>condition added</strong> · <a href="https://github.com/dotnet/maui/commit/fc467b1c507385c2d97ccd963abbc55bff566062"><code>fc467b1</code></a></summary>

---

##  Final Recommendation: APPROVE

### Summary
PR #31248 fixes issue #19541 where SwipeView stays open on iOS after CollectionView content is updated. The fix is minimal, targeted, and correctly platform-gated. Gate verification confirmed tests fail without the fix and pass with it. Four independent fix alternatives were explored via try-fix - all passed but the PR's approach is the simplest and most appropriate.

### Root Cause
On iOS/MacCatalyst, `CollectionView` **reuses cells** (UITableView/UICollectionView cell reuse pattern). When `ItemsSource` is refreshed, existing cells are recycled and their `BindingContext` is updated - but the `SwipeView._isOpen` state from the previous binding persists visually. On Android and Windows, new cells are created on rebind, so `_isOpen` naturally starts as `false`.

### Fix Quality
The fix is well-designed:
- **Minimal**: 12 lines added to a single file (`SwipeView.cs`)
- **Platform-gated**: Uses `#if IOS || MACCATALYST` - correct for this iOS-specific cell reuse behavior
- **Guarded**: `if (_isOpen)` prevents unnecessary `RequestClose` calls on already-closed SwipeViews
- **Uses correct API**: `((ISwipeView)this).RequestClose(new SwipeViewCloseRequest(false))` - uses the established interface with `animated: false` (no animation needed during data refresh)
- **Well-commented**: Explains the platform difference (cell reuse vs new cells)

### Test Quality
- **HostApp page** (`Issue19541.cs`): Creates a CollectionView with SwipeViews, with Open and Refresh buttons. Correctly uses `AutomationId` for test automation.
- **NUnit test**: Opens SwipeView, triggers refresh, verifies via screenshot. Simple and direct.
- **Platform guard**: `#if TEST_FAILS_ON_WINDOWS` correctly excludes Windows (iOS-specific bug).
- **Snapshots**: Added for iOS, Android, and Mac.

### Reviewer Thread Status
One review thread from  asking for a compilation directive for iOS/Catalyst -  **already addressed** by the author. Thread is marked unresolved in the UI but the code change was made.

### Fix Comparison
| Approach | Complexity | Files | Verdict |
|----------|------------|-------|---------|
| PRs  Minimal | 1 (+12 lines Best |) | fix | 
| Attempt 1 (IsOpen mapper) | High | 3 + 2 API files | New public API required |
| Attempt 2 (unconditional reset) | Low | 1 | Fires even when not open |
| Attempt 3 (ItemsSource listener) | Medium | 1 | Couples to CollectionView |
| Attempt 4 (Parent=null) | Low | 1 | Too broad |

</details>

</details>

---

</details>



<details>
<summary>📋 <strong>Expand PR Finalization Review</strong></summary>

---

<details>

<summary><b>Title: ✅ Good</b></summary>

<br>

**Current:** `[iOS] Fix SwipeView stays open on iOS after updating content`
</details>

<details>

<summary><b>Description: ✅ Good</b></summary>

<br>

Description needs updates. See details below.
### ✨ Suggested PR Description


> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

### Root Cause

On iOS and Mac Catalyst, `UICollectionView` **reuses cells** when the `ItemsSource` is refreshed. This means the same `SwipeView` instance is rebound with a new `BindingContext` while still in the open state (`_isOpen = true`). The swipe items remain visually open even though the underlying data has changed.

On Android and Windows, new cells are created for each item on rebind, so `_isOpen` is always `false` for new instances — no issue.

### Description of Change

In `SwipeView.OnBindingContextChanged()`, added an iOS/MacCatalyst-only check: if the SwipeView is currently open when the `BindingContext` changes, `RequestClose` is called (without animation) to reset the open state.

```csharp
#if IOS || MACCATALYST
    if (_isOpen)
    {
        ((ISwipeView)this).RequestClose(new SwipeViewCloseRequest(false));
    }
#endif
```

This closes the stale swipe state before the view is reused with the new binding context.

### Key Technical Details

- **`_isOpen`** — internal field on `SwipeView` tracking open state
- **`ISwipeView.RequestClose`** — dispatches to the platform handler via `Handler?.Invoke(...)`, so it's safe to call even if no handler is attached
- **`SwipeViewCloseRequest(false)`** — `false` = no animation, appropriate for a reset during rebind
- The `#if IOS || MACCATALYST` conditional is consistent with existing patterns in `SwipeView.cs`

### Issues Fixed

Fixes #19541

### Platforms Tested

- [x] Android
- [x] Windows
- [x] iOS
- [x] Mac
</details>

<details>

<summary><b>Code Review: ✅ Passed</b></summary>

<br>

# Code Review — PR #31248

## 🟡 Suggestions

### 1. Inconsistent field naming in HostApp test (`Issue19541.cs`)

**File:** `src/Controls/tests/TestCases.HostApp/Issues/Issue19541.cs`

`_swipeView` uses the conventional underscore prefix, but `collectionView` does not.

```csharp
// Current (inconsistent)
SwipeView _swipeView;
CollectionView collectionView;

// Suggested (consistent)
SwipeView _swipeView;
CollectionView _collectionView;
```

This is a minor style issue and won't affect functionality.

---

### 2. Behavioral edge case: closes on *any* BindingContext change

**File:** `src/Controls/src/Core/SwipeView/SwipeView.cs`

The fix closes the SwipeView whenever `_isOpen && BindingContext changes` on iOS/MacCatalyst — not just when inside a CollectionView. This means if a user manually updates the `BindingContext` of a SwipeView that the user intentionally opened (outside of a list), it will be silently closed.

This is a reasonable tradeoff: the reported scenario (CollectionView cell reuse) is the primary and known use case. However, it's a behavior change worth documenting.

**Recommendation:** The code comment already explains the reasoning well. No code change needed, but the PR description should mention this tradeoff (addressed in `recommended-description.md`).

---

### 3. `swipeView_Loaded` captures only the first SwipeView

**File:** `src/Controls/tests/TestCases.HostApp/Issues/Issue19541.cs`

```csharp
void swipeView_Loaded(object sender, EventArgs e)
{
    if (_swipeView is null)
        _swipeView = (SwipeView)sender;
}
```

This assigns `_swipeView` only once (the first to load). The "Open" button then opens specifically that first cell's SwipeView, which is fine for the test scenario. The test verifies that after refresh, the SwipeView is closed — the exact cell tracked is sufficient.

No action needed, but this is worth noting for future test modifications.

---

## ✅ Looks Good

- **Fix scope is correct:** `#if IOS || MACCATALYST` is the right conditional. The existing file already uses `#if IOS` for platform-specific logic, so this is consistent.
- **`RequestClose` is null-safe:** Uses `Handler?.Invoke(...)` so it won't throw if called before the handler is attached.
- **No animation on close (`false`):** Appropriate for a background reset during cell reuse — no visual flash.
- **UI test covers the scenario:** The test opens a SwipeView, refreshes the CollectionView, and verifies via screenshot — directly reproducing the reported bug.
- **Test excluded from Windows (`#if TEST_FAILS_ON_WINDOWS`):** Correct, as the issue is iOS-specific.
- **Snapshot baselines added** for iOS, Android, and Mac — the test infrastructure is complete.
</details>

---

</details>

**Review by kubaflo — APPROVED** [review]
